src/smallstring.jl

```diff
@@ -1,66 +1,86 @@
+export convert,
+       push,
+       setindex
 
 const IntChar = UInt8
 const IntSmallString = UInt64
 const maxTagLength = 8
 const SmallStringStorage = SVector{maxTagLength,IntChar}
 const MSmallStringStorage = MVector{maxTagLength,IntChar} # Mutable SmallString storage
 
+function initSmallStringStore()
+  return SmallStringStorage(ntuple(_->IntChar(0),Val(length(SmallStringStorage))))
+end
+
 struct SmallString
   data::SmallStringStorage
-  length::Int
-  SmallString(sv::SmallStringStorage,l::Int) = new(sv,l)
+
+  SmallString(sv::SmallStringStorage) = new(sv)
+
   function SmallString()
-    sv = SmallStringStorage(ntuple(_ -> IntChar(0),Val(length(SmallStringStorage))))
-    return new(sv, 0)
+    sv = initSmallStringStore()
+    return new(sv)
   end
-end
 
-function SmallString(sv::SmallStringStorage)
-  sv[1] == IntChar(0) && return SmallString(sv,0)
-  len = 1
-  while len < length(SmallStringStorage) && @inbounds sv[len+1] ≠ IntChar(0)
-    len += 1
+  function SmallString(str::String)
+    sv = initSmallStringStore()
```

Comment:
Could this function be done by creating an `MSmallStringStorage` and then filling it with the values with `setindex!`, then converting to a `SmallStringStorage`/`SmallString`? That should still avoid allocations, and I believe I have seen that pattern was a bit faster.

Reply:
Good feedback! Ok hopefully improved in the latest commits. Please take a look.